qiskit/qasm2/__init__.py

```diff
@@ -0,0 +1,523 @@
+letter-of-the-spec mode with ``strict=True``.
+
+Errors
+======
+
+This module defines a generic error type that derives from :exc:`.QiskitError` that can be used as a
+catch when you care about failures emitted by the interoperation layer specifically.
+
+.. autoexception:: QASM2Error
+
+In cases where the lexer or parser fails due to an invalid OpenQASM 2 file, the conversion functions
+will raise a more specific error with a message explaining what the failure is, and where in the
+file it occurred.
+
+.. autoexception:: QASM2ParseError
+
+.. _qasm2-examples:
+
+Examples
+========
+
+The most basic example is to use :func:`loads` to import an OpenQASM 2 program in a Python string
+into a :class:`.QuantumCircuit`:
+
+.. code-block:: python
+
+    import qiskit.qasm2
+    program = '''
+        OPENQASM 2.0;
+        include "qelib1.inc";
+        qreg q[2];
+        creg c[2];
+
+        h q[0];
+        cx q[0], q[1];
+
+        measure q -> c;
+    '''
+    qiskit.qasm2.loads(program).draw()
+
+.. code-block:: text
+
+         ┌───┐     ┌─┐
+    q_0: ┤ H ├──■──┤M├───
+         └───┘┌─┴─┐└╥┘┌─┐
+    q_1: ─────┤ X ├─╫─┤M├
+              └───┘ ║ └╥┘
+    c: 2/═══════════╩══╩═
+                    0  1
+
+You can achieve the same thing if the program is stored in a file by using :func:`load` instead,
+passing the filename as an argument:
+
+.. code-block:: python
+
+    import qiskit.qasm2
+    circuit = qiskit.qasm2.load("myfile.qasm")
+
+OpenQASM 2 files can include other OpenQASM 2 files via the ``include`` statement.  You can
+influence the search path used for finding these files with the ``include_path`` argument to both
+:func:`load` and :func:`loads`.  By default, only the current working directory is searched.
+
+.. code-block:: python
+
+    import qiskit.qasm2
+    program = '''
+        include "other.qasm";
+        // ... and so on
+    '''
+    circuit = qiskit.qasm2.loads(program, include_path=("/path/to/a", "/path/to/b", "."))
+
+For :func:`load` only, there is an extra argument ``include_input_directory``, which can be used to
+either *append*, *prepend* or ignore (``None``) the directory of the loaded file in the include
+path.  By default, this directory is appended to the search path, so it is tried last, but you can
+change this.
+
+.. code-block:: python
+
+    import qiskit.qasm2
+    paths = ["./subdirectory/a.qasm", "/path/to/b.qasm", "~/my.qasm"]
+    # Search the directory of each file before other parts of the include path.
+    circuits = [qiskit.qasm2.load(path, include_input_directory="prepend") for path in paths]
+    # Override the include path, and don't search the directory of each file unless it's in the
+    # absolute path list.
+    circuits = [
+        qiskit.qasm2.load(
+            path,
+            include_path=("/usr/include/qasm", "~/qasm/include"),
+            include_input_directory=None,
+        )
+        for path in paths
```

Comment:
the term `path` is a bit overloaded in this context, I would go this way:

```suggestion
    qasm_files = ["./subdirectory/a.qasm", "/path/to/b.qasm", "~/my.qasm"]
    # Search the directory of each file before other parts of the include path.
    circuits = [qiskit.qasm2.load(qasm_file, include_input_directory="prepend") for qasm_file in qasm_files]
    # Override the include path, and don't search the directory of each file unless it's in the
    # absolute path list.
    circuits = [
        qiskit.qasm2.load(
            qasm_file,
            include_path=("/usr/include/qasm", "~/qasm/include"),
            include_input_directory=None,
        )
        for qasm_file in qasm_files
```

Reply:
Done (something similar) in 6048257b3.